pygbm/loss.py

```diff
@@ -0,0 +1,104 @@
+from abc import ABC
+
+import numpy as np
+from numba import njit, prange
+import numba
+
+
+@njit
+def _get_threads_chunks(total_size):
+    # Divide [0, total_size - 1] into n_threads contiguous regions, and
+    # returns the starts and ends of each region. Used to simulate a 'static'
+    # scheduling.
+    n_threads = numba.config.NUMBA_DEFAULT_NUM_THREADS
+    sizes = np.full(n_threads, total_size // n_threads, dtype=np.int32)
+    sizes[:total_size % n_threads] += 1
+    starts = np.zeros(n_threads, dtype=np.int32)
+    starts[1:] = np.cumsum(sizes[:-1])
+    ends = starts + sizes
+
+    return starts, ends, n_threads
+
+
+@njit(fastmath=True)
+def _sigmoid(x):
+    # custom sigmoid because we cannot use that of scipy with numba
+    return 1 / (1 + np.exp(-x))
+
+
+class Loss(ABC):
+
+    def init_gradients_and_hessians(self, n_samples):
+        gradients = np.empty(shape=n_samples, dtype=np.float32)
+        if self.hessian_is_constant:
+            hessians = np.ones(shape=1, dtype=np.float32)
+        else:
+            hessians = np.empty(shape=n_samples, dtype=np.float32)
+
+        return gradients, hessians
+
+
+class LeastSquares(Loss):
+
+    hessian_is_constant = True
+
+    def __call__(self, y_true, y_pred, average=True):
+        loss = np.power(y_true - y_pred, 2)
+        return loss.mean() if average else loss
+
+    def update_gradients_and_hessians(self, gradients, hessians, y_true,
+                                      y_pred):
+        return _update_gradients_least_squares(gradients, y_true, y_pred)
+
+    def get_gradients(self, y_true, y_pred):
```

Comment:
@ogrisel As we're now using `update_gradients_and_hessians` and as it's inplace, I removed the `get_hessians` methods and only kept `get_gradients` because it's still used in the tests.

Let me know if you think we could get rid of it? Maybe `test_numerical_gradients` is enough?

Reply:
Maybe it's possible to remove this method from the loss but to rewrite the other test to add an local function in the test that needs `get_gradients`  to actually rely on `update_gradients_and_hessians` under the hood?